src/main/java/com/conveyal/gtfs/loader/Table.java

```diff
@@ -746,37 +747,38 @@ public void checkReferencesAndUniqueness(String keyValue, int lineNumber, Field
             }
         }
 
+        // These hold references to the set of IDs to check for duplicates and the ID to check. These depend on
+        // whether an order field is part of the "unique ID."
+        Set<String> listOfUniqueIds = referenceTracker.transitIds;
+        String uniqueId = transitId;
+
+        // There is no need to check for duplicate IDs if the field is a foreign reference.
+        if (field.isForeignReference()) return;
+        // Next, check that the ID is table-unique.
         if (field.name.equals(uniqueKeyField)) {
-            // These hold references to the set of IDs to check for duplicates and the ID to check. These depend on
-            // whether an order field is part of the "unique ID."
-            String transitId = String.join(":", getKeyFieldName(), keyValue);
-            Set<String> listOfUniqueIds;
-            String uniqueId;
             // Check for duplicate IDs and store entity-scoped IDs for referential integrity check
             if (isOrderField) {
                 // Check duplicate reference in set of field-scoped id:sequence (e.g., stop_sequence:12345:2)
                 // This should not be scoped by key field because there may be conflicts (e.g., with trip_id="12345:2"
                 listOfUniqueIds = referenceTracker.transitIdsWithSequence;
                 uniqueId = String.join(":", field.name, keyValue, string);
-                if (!field.isForeignReference() && !referenceTracker.transitIds.contains(transitId)) {
-                    // Only add ID if the field is not a foreign reference (e.g., adds shape_id for shapes.txt,
-                    // but skips trip_id in stop_times.txt)
-                    referenceTracker.transitIds.add(transitId);
-                }
-            } else {
-                listOfUniqueIds = referenceTracker.transitIds;
-                uniqueId = transitId;
             }
             // Add ID and check duplicate reference in entity-scoped IDs (e.g., stop_id:12345)
-            boolean isDuplicate = !listOfUniqueIds.add(uniqueId);
-            if (isDuplicate) {
+            boolean valueAlreadyExists = !listOfUniqueIds.add(uniqueId);
+            if (!this.name.equals("calendar_dates") && valueAlreadyExists) {
```

Comment:
It seems strange to have a hard coded single exception here. This code is already inside a conditional `field.name.equals(uniqueKeyField)`. Do we actually need to flag this field as the "unique key field" in `calendar_dates` when it's not unique? Is it even a good idea to do so?

Reply:
I think I see what you mean.  `service_id` should not be the "unique key field" for `calendar_dates` in the first place.  So when we set the value of `uniqueKeyField`, we could do something like the following: 
```
String uniqueKeyField = this.name.equals("calendar_dates") 
? null
: orderField != null ? orderField : keyField;
```

This still hard codes the calendar_dates exception, but the meaning is a little clearer.

Reply:
That's basically what I mean, but I was thinking one level even further up. `Table.getKeyFieldName` always returns the first field in the table, but several tables don't have a unique key (nor a compound key using a sequence number). There are several tables that will create this same situation. Scanning through the table definitions, I see transfers, frequencies, feed_info, and fare_rules where the first field does not serve as a unique key, neither alone nor in combination with a sequence number. For those tables, I'd say getKeyField should return null. We might need another fluent method to set the key field (or de-set it) on these tables.

Reply:
Ah, right.  Yes, that makes sense to me.